database/scrapers/event_calendar.ts

```diff
@@ -0,0 +1,227 @@
+    }
+    this.events = eventArray;
+  }
+
+  public *getLastEvent(): Generator<GoogleCalendarEventDto, void, undefined> {
+    const validEvents = this.events
+      .map((e) => this.extractEvent(e))
+      .filter((e) => e !== undefined)
+      .sort((a, b) => {
+        const diff = a.endTime.diff(b.endTime).milliseconds;
+        if (diff !== 0) {
+          return diff;
+        }
+        return a.startTime.diff(b.startTime).milliseconds;
+      }); //sorted from earliest to last, with one most recent/in the future being last
+    for (let i = validEvents.length - 1; i >= 0; i--) {
+      yield validEvents[i];
+    }
+  }
+
+  //null - none left, undefined - malformed event
+  private extractEvent(
+    eventBlock: ICSObject,
+  ): GoogleCalendarEventDto | undefined {
+    if (Array.isArray(eventBlock)) {
+      this.task.update(
+        "Invalid calendar event - block is not an ICSObject. Skipping...",
+      );
+      return undefined;
+    }
+    return this.parseEvent(eventBlock);
+  }
+
+  private static asValidDateTime(
+    dateValue: ICSValue | undefined,
+  ): DateTime<true> | null {
+    if (dateValue === undefined || Array.isArray(dateValue)) {
+      return null;
+    }
+    if (typeof dateValue === "string") {
+      //timestamp: dateValue = 20180424T00:13:21
+      const dateTime = DateTime.fromISO(dateValue);
+      if (dateTime.isValid) {
+        return dateTime;
+      }
+      return null;
+    }
+    const nestedDate = dateValue.VALUE; //full day date:  dateValue = { "VALUE": "DATE:20180421" }
+    if (
+      nestedDate === undefined ||
+      typeof nestedDate !== "string" ||
+      nestedDate.length !== 13
+    ) {
+      return null;
+    }
+    const dateTime = DateTime.fromISO(nestedDate.substring(5));
+    if (dateTime.isValid) {
+      return dateTime;
+    }
+    return null;
+  }
+
+  private static asString(value: ICSValue | undefined): string | null {
+    if (value === undefined || typeof value !== "string") {
+      return null;
+    }
+    return value;
+  }
+
+  private parseEvent(eventBlock: ICSRoot): GoogleCalendarEventDto | undefined {
+    const startTime = CalendarParser.asValidDateTime(
+      eventBlock[FIELD_LABELS.startDate],
+    );
+    const endTime = CalendarParser.asValidDateTime(
+      eventBlock[FIELD_LABELS.endDate],
+    );
+    if (startTime === null || endTime === null) {
+      this.task.update(
+        "Invalid calendar event, missing date labels. Skipping...",
+      );
+      return undefined;
+    }
+    const googleCalId = CalendarParser.asString(
+      eventBlock[FIELD_LABELS.googleCalId],
+    );
+    if (googleCalId === null) {
+      this.task.update(
+        "Invalid calendar event, missing googleCalId. Skipping...",
+      );
+      return undefined;
+    }
+    const name = CalendarParser.asString(eventBlock[FIELD_LABELS.name]);
+    if (name === null) {
+      this.task.update("Invalid calendar event, missing name. Skipping...");
+      return undefined;
+    }
```

Comment:
would be nice to still have the event indexes in these warnings
also `this.logger.warn` is a more better function for logging issues than `task.update`.
or the `logger` function you pass into the constructor...